### PR TITLE
nRF Desktop Failsafe module: migration to the Zephyr HWinfo driver and activation on the nRF54H20 DK

### PR DESCRIPTION
**Pull request by kapi-no**
Depends on:

~https://github.com/nrfconnect/sdk-nrf/pull/19028~

Issues:
- ~Watchdog reset reason bit is always set for the nRF54H20 DK regardless whether the reset is caused by the reset pin (`RESET_PIN`: `0x00000011`) or software reset (`RESET_SOFTWARE`: `0x00000012`).~

Out standing items:
- [x] Changelog and the documentation update for the Failsafe module.

### CODE REVIEW

File: applications/nrf_desktop/src/modules/failsafe.c [hunk 1]
```diff
@@ -15,14 +15,24 @@
...
 
-	uint32_t reas = nrfx_reset_reason_get();
+	err = hwinfo_get_reset_cause(&reas);
+	if (err) {
+		return err;
+	}
+
+	*ret = ((reas & mask) != 0);
+
+	LOG_INF("Reset reason (0x%08X) trigger %s for Failsafe erase operation",
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/failsafe.c:32]: Can we make it a bit shorter: `"Reset reason (0x%08X), trigger %s`? Log would be prefixed with module name by the logger anyway

File: applications/nrf_desktop/src/modules/failsafe.c [hunk 2]
```diff
@@ -43,9 +53,11 @@ static void failsafe_erase(void)
 	}
 }
 
-static void failsafe_clear(void)
+static int failsafe_clear(void)
 {
-	nrfx_reset_reason_clear(nrfx_reset_reason_get());
+	int err = hwinfo_clear_reset_cause();
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/failsafe.c:58]: `return hwinfo_clear_reset_cause();` - for simplicity?

File: applications/nrf_desktop/src/modules/failsafe.c [hunk 3]
```diff
@@ -55,17 +67,27 @@ static bool app_event_handler(const struct app_event_header *aeh)
...
-			if (failsafe_check()) {
+			err = failsafe_check(&check_result);
+			if (!err && check_result) {
 				failsafe_erase();
 			}
-			failsafe_clear();
 
-			module_set_state(MODULE_STATE_READY);
+			if (!err) {
+				err = failsafe_clear();
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/failsafe.c:83]: Consider avoiding clearing the reset reason if `failsafe_erase` call fails.

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: applications/nrf_desktop/src/modules/failsafe.c (hunk 3 above)
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/failsafe.c:83]: This worked a bit differently in the previous implementation, where we cleared the reset reason unconditionally. Are you sure about that?

Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/failsafe.c:83]: I am aware that it used to work differently.
I am wondering a bit which approach would be better here: if we desist from clearing the reset reason on settings erase error, we give failsafe module another chance to erase settings which could be safer.

File: applications/nrf_desktop/src/modules/failsafe.c [hunk 4]
```diff
@@ -15,14 +15,24 @@
 LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_FAILSAFE_LOG_LEVEL);
 
 
-static bool failsafe_check(void)
+static bool failsafe_check(bool *ret)
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/failsafe.c:18]: I prefer failure detected as the settings erase may be one of many actions in the future that this module needs to perform on failure detection

File: applications/nrf_desktop/src/modules/failsafe.c (hunk 2 above)
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/failsafe.c:58]: I added some additional error logs

File: applications/nrf_desktop/src/modules/failsafe.c [hunk 5]
```diff
@@ -55,17 +67,27 @@ static bool app_event_handler(const struct app_event_header *aeh)
 				cast_module_state_event(aeh);
 
 		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+			int err;
+			bool check_result;
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/failsafe.c:71]: `failure_detected` for consistency

**Comment by kapi-no** [comment]
@MarekPieta, addressed your comments

**Comment by kapi-no** [comment]
The Watchdog issue should be fixed after addressing the comments from the Zephyr upstream PR that adds reset reason support for nRF54H20 SoC in the Zephyr HWinfo driver:

https://github.com/zephyrproject-rtos/zephyr/pull/81751

File: applications/nrf_desktop/src/modules/failsafe.c [hunk 6]
```diff
@@ -15,17 +15,28 @@
 LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_FAILSAFE_LOG_LEVEL);
 
 
-static bool failsafe_check(void)
+static bool failsafe_check(bool *failure_detected)
```
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/failsafe.c:18]: `static int`

**Review by MarekPieta — APPROVED** [review]

File: applications/nrf_desktop/src/modules/failsafe.c (hunk 6 above)
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/failsafe.c]: thanks, fixed

**Review by MarekPieta — APPROVED** [review]

**Comment by kapi-no** [comment]
The issue is resolved as the HW info driver fixes are merged as part of the following PR:

https://github.com/nrfconnect/sdk-zephyr/pull/2385

**Review by MarekPieta — APPROVED** [review]

**Review by mkapala-nordic — APPROVED** [review]

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 7]
```diff
@@ -198,7 +198,11 @@ nRF5340 Audio
 nRF Desktop
 -----------
 
-|no_changes_yet_note|
+* Updated:
+
+  * The :ref:`nrf_desktop_failsafe` to use the Zephyr :ref:`zephyr:hwinfo_api` driver for getting and clearing the reset reason information (see the :c:func:`hwinfo_get_reset_cause` and :c:func:`hwinfo_clear_reset_cause` function).
+    The Zephyr :ref:`zephyr:hwinfo_api` driver replaces the nrfx reset reason helper depedency (see the :c:func:`nrfx_reset_reason_get` and :c:func:`nrfx_reset_reason_clear` function).
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:204]:
```suggestion
    The Zephyr :ref:`zephyr:hwinfo_api` driver replaces the nrfx reset reason helper dependency (see the :c:func:`nrfx_reset_reason_get` and :c:func:`nrfx_reset_reason_clear` functions).
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 8]
```diff
@@ -198,7 +198,11 @@ nRF5340 Audio
 nRF Desktop
 -----------
 
-|no_changes_yet_note|
+* Updated:
+
+  * The :ref:`nrf_desktop_failsafe` to use the Zephyr :ref:`zephyr:hwinfo_api` driver for getting and clearing the reset reason information (see the :c:func:`hwinfo_get_reset_cause` and :c:func:`hwinfo_clear_reset_cause` function).
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:203]:
```suggestion
  * The :ref:`nrf_desktop_failsafe` to use the Zephyr :ref:`zephyr:hwinfo_api` driver for getting and clearing the reset reason information (see the :c:func:`hwinfo_get_reset_cause` and :c:func:`hwinfo_clear_reset_cause` functions).
```

File: applications/nrf_desktop/src/modules/failsafe.c [hunk 9]
```diff
@@ -15,17 +15,28 @@
...
 
 
-static bool failsafe_check(void)
+static int failsafe_check(bool *failure_detected)
 {
-	uint32_t mask = NRFX_RESET_REASON_DOG_MASK |
-			NRFX_RESET_REASON_LOCKUP_MASK;
+	int err;
+	uint32_t reas = 0;
+	uint32_t mask = RESET_WATCHDOG |
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply applications/nrf_desktop/src/modules/failsafe.c]: improved

**Review by divipillai — APPROVED** [review]

**Comment by kapi-no** [comment]
Rebased due to the CI failures. No file diff has changed (pure rebase).

**Comment by kapi-no** [comment]
@nordicjm, could you review the build system part?

**Review by nordicjm — APPROVED** [review]

**Comment by kapi-no** [comment]
Rebased due to the CI failures. No file diff has changed (pure rebase).

**Pull request merged**